### PR TITLE
New changes

### PR DESCRIPTION
**Pull request by chaim-f**
This version is based on the solution downloaded from the master file at 7pm November 10. things changed in this version:

1. Deleted individual drivers
2. In Cards.cpp Added stream overload
3. In map.cpp and .h fixed constructor and added copy constructors and destructors to Territory and Contenent classes
4. In player class fixed the addcardToHandOfCards function and fixed some of the constructor
5. Also in the player class changed playerID from const char* to string (therefore I also had to change this in the DriverSelector)
6. In the MapLoader class added a parameter constructor
Reasons:

1. due to the driverSelector we don't need the individual drivers anymore
2. Was missing
3. was missing or incomplete
4. was missing
5. as const char needs to be initialized when declared
6. So I do not need to run the functions set FileName, storeContinents, storeTerritories, and storeTerritoriesWithBorders it will do it automaticly
To Do:

Map.cpp needs setters for the Territory and Contenent Classes, line 31 needs someone who understands Lists to delete the list (maybe have to itterate accross the list I don't know) and Lines 52 and 69 needs someone who understands lists to assign the values from the original to the copyer, and line 300 needs to assign the default value to the destination (I am not sure if it is 0 or -1 or even if it makes a diffrence as I don't know if destination of the Territory is used at all).
Player.cpp the copy constructors and equals operator currently copy the same value need a tempThing =new type(*orig.thing) and then needs the tempThing to be pushed. Also functions such as addMyTerritory currently only prints that it did something does not actually do it.
Also in Player.cpp need to fix the destructor.
In Cards.cpp change from const string to enum (I am too lazy to do so as it was just sugested not mandatory).
Questions

1. Map.cpp line 330 why does it used two ways of checking if destination>=1 maybe it was supposed to check anther variable?

### CODE REVIEW

File: Warzone/Map.h [hunk 1]
```diff
@@ -1,72 +1,82 @@
...
+	 //std::vector <Territory*> MapterritoryVector;
+	 std::vector <Territory*> MapContinentVector;
+	 bool validate();
+	 string getMapName();
+	 void setMapName(string mName);
+ private:
+	 string* mapName;
+	 bool* validMap;
+ };
+
```
Comment (tramyardg) [inline Warzone/Map.h:82]: Why are you rewriting the whole Map class, if you don't understand please ask. This class is used in my part.

File: Warzone/MapLoader.cpp [hunk 2]
```diff
@@ -1,239 +1,30 @@
...
+	b.getTerritoryName();
+	//c1.setInContinent(3);
+	a.addAdjacent(&b);
+	a.addAdjacent(&c);
+	cout<<a.toString();
+	cout << c1.toString();
+	cout << "enter any char and then enter to exit";
+	char d1;
+	cin >> d1;
+	return 0;
```
Comment (tramyardg) [inline Warzone/MapLoader.cpp:29]: Same here, you just ignored/disregarded everything I and Jeremy worked for from asg 1, it is already working as it is, there is no need to make a big changes. I am also relying on this on my part. Please work with what we have don't make big changes.

**Review by tramyardg — CHANGES_REQUESTED** [review]

**Comment by tramyardg** [comment]
To answer your question: "ap.cpp line 330 why does it used two ways of checking if destination>=1 maybe it was supposed to check anther variable?"
in a [`.map`](http://domination.sourceforge.net/getmaps.shtml) file territory index **always starts at 1**, for example
```
; map: europe.map
; map made with the map maker
; yura.net Risk 1.0.9.5

[files]
...

[continents]
Scandinavia 5 #fff780
....

[countries]
1 s1 1 66 28
2 s2 1 384 39
...
```
Now what if I have a custom map that starts at territory src 0, for example, `Territory(0, 1, "Quebec")`. So this is just validation checks required to do that.

File: Warzone/MapLoader.cpp (hunk 2 above)
Comment (chaim-f) [reply Warzone/MapLoader.cpp:29]: @tramyardg  there is some of your code I do not understand 
In the map class 
1. you have a vector of pointers, where are the pointers deleted?
2. In the map destructor why is there the brackets by delete[] adjList, adjList is a pointer to a list of Territories we can just delete adjList.
3. The copy constructor and the = operator should be coping from m or t, why is it assuming m and t are the default values, what if we called functions to add things to it then called the constructor or equals operator?
4. Are we relying on default initialization for half of the constructors and what value and what happens if the constructor which does not have a destination (Territory(int src, string name, Continent c);) was used and then the command of addEdge from map was called on that edge (do we have any function to catch that and prevent it from crashing?
5.why is the getters (getSource() and getDestination()) declared as friend and not declared in the function (continent::getSource()...)?
6. if have constructors that do not take and initialize every value why do we not have setters so the values can be initialized later on?
7. And while truthfully we do not need to overload the equal operator for the continent and  Territory classes for the sake of the assignment we need them same for the copy constructors.

For the MapLoader 
1. why is the destructor empty, (where are you deleting the pointers in the vectors)?
2. why is the copy constructor and = operator making the file name "maps/canada.map" what if the file we are copying from has a diffrent file name?

Although this was not done by you I also have questions on this classes that maybe you can answer.
In the Player class.
1. why is the comment to do in lines 35-40 still there was have all of those in the class.
2. what is the purpose of this variable Player * thisPlayerRef = nullptr;
3. why is the playerID a const as if it is a const it must be initialized when it is declared?
4.  In the copy constructor when you have something like this (//Copy territories
	for (ptrTerritory = anotherPlayer.myTerritories->begin(); ptrTerritory != anotherPlayer.myTerritories->end(); ++ptrTerritory) {
		myTerritories->push_back(*ptrTerritory);
})
is this a deep or shallow copy?
5. why are there both issueOrder and addOrder when they seem to be doing the same thing?
6. why is the implementation of addMyTerritory and addTargetTerritory commented out?
7. why is there no removeMyTerritory and removeTargetTerritory functions?

I will admit that the cards class also has errors and I am in middle of fixing them, also my thoughts were anything that we anyway need to fix we should also simplify knowing from the FAQ that any non self defined functions do not need to be pointers (such as int, vector<>, and list<>.

Comment (chaim-f) [reply Warzone/MapLoader.cpp:29]: @tramyardg I will just assume you guys did correct and delete this comment as well as the earlier ones.

**Pull request closed without merging**